feat(movies): added findById and list to /movies endpoint with tests

wcthompson opened the pull request:
What: Added GET route to `/movies` endpoint and tests

Why: Not a very useful api if you can only store!

Details:
- can get a single movie by id (`/movies/{id}`)
- get a filtered list via query params (`/movies?title='Spirited Away'`, `/movies?release_year='2001'`)
- validation, controller and integration tests using a movie factory

charlesbjohnson commented on lib/plugins/features/movies/controller.js:
```diff
@@ -8,3 +8,24 @@ exports.create = (payload) => {
+    } else if ('release_year' in query) {
+      qb.where('release_year', query.release_year);
+    }
+  })
```
Should these conditions be exclusive? This probably wouldn't allow me to lookup the [Bad Boys](http://www.imdb.com/title/tt0085210/) that I care about, and not that [other one](http://www.imdb.com/title/tt0112442/) that no one has ever heard of before.

charlesbjohnson commented on lib/plugins/features/movies/controller.js:
```diff
@@ -8,3 +8,24 @@ exports.create = (payload) => {
+  return new Movie({ id: query }).fetch({ require: true })
+  .catch(Movie.NotFoundError, () => {
+    throw new Error('Movie not found');
+  })
```
Do you think the `catch` is needed? Is the `Movie.NotFoundError`'s message a lot worse than the one we are providing here?

charlesbjohnson commented on lib/plugins/features/movies/index.js:
```diff
@@ -1,7 +1,8 @@
 const Controller     = require('./controller');
-const MovieValidator = require('../../../validators/movie');
+const MovieCreateValidator = require('../../../validators/movies/create');
+const MovieListValidator = require('../../../validators/movies/list');
```
nit: align `=`

charlesbjohnson commented on lib/validators/movies/create.js:
```diff
@@ -4,5 +4,5 @@ const Joi = require('joi');
 module.exports = Joi.object().keys({
   title: Joi.string().min(1).max(255).required(),
-  release_year: Joi.number().integer().min(1878).max(9999).optional()
+  release_year: Joi.number().min(1878).max(9999).optional()
```
I think we still want/need `integer`, no?

charlesbjohnson commented on test/factories/movie.js:
```diff
@@ -0,0 +1,8 @@
+
+module.exports = Factory.define('movie')
+  .sequence('id')
+  .attr('title', (title) => `${title}`)
```
Do we need the function generator? It seems redundant.

charlesbjohnson commented on test/factories/movie.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = Factory.define('movie')
+  .sequence('id')
+  .attr('title', (title) => `${title}`)
+  .attr('release_year', '');
```
Since `release_year` is a number, shouldn't the default value be as well?

charlesbjohnson commented on test/plugins/features/movies/index.test.js:
```diff
@@ -1,6 +1,12 @@
 'use strict'
 
 const Movies = require('../../../../lib/server');
+const Knex       = require('../../../../lib/libraries/knex');
```
nit: align `=`

charlesbjohnson commented:
Done with first pass. Ping for re-review.

wcthompson commented:
@charlesbjohnson thanks! fixed up and added and several controller tests to cover missing cases (find multiple movies with same release year, same name, specific name+year).

charlesbjohnson commented on lib/plugins/features/movies/controller.js:
```diff
@@ -8,3 +8,22 @@ exports.create = (payload) => {
+    if ('title' in query) {
+      qb.where('title', query.title);
+    }
+    if ('release_year' in query) {
```
optional: This works fine as is, so no changes are necessary. For some context, though, I tend not to use `in`. A better alternative for just general checking if a key exists, IMO, is `hasOwnProperty` since it doesn't travel up the prototype chain. Another alternative that is more conventional to Lob is to check the truthiness of the property, ie.
```js
if (query.title) {
  ...
}
```

since it would automatically rule out the empty string, `undefined`, or `null`.

charlesbjohnson commented on test/plugins/features/movies/controller.test.js:
```diff
@@ -1,8 +1,16 @@
+const spirited_away = MovieFactory.build({ title: 'Spirited Away', release_year: 2001 });
+const hackers       = MovieFactory.build({ title: 'Hackers ', release_year: 1995 });
+const bad_boys_1    = MovieFactory.build({ title: 'Bad Boys', release_year: 1983 });
+const bad_boys_2    = MovieFactory.build({ title: 'Bad Boys', release_year: 1995 });
```
nit: prefer camelCasing in this case

charlesbjohnson commented on test/plugins/features/movies/controller.test.js:
```diff
@@ -1,8 +1,16 @@
+const MovieFactory = require('../../../factories/movie');
+
+const spirited_away = MovieFactory.build({ title: 'Spirited Away', release_year: 2001 });
+const hackers       = MovieFactory.build({ title: 'Hackers ', release_year: 1995 });
```
nit: extra space in `title`

charlesbjohnson commented:
LGTM
